src/System.Private.CoreLib/shared/System/Number.Formatting.cs

```diff
@@ -387,61 +398,28 @@ public static bool TryFormatDouble(double value, ReadOnlySpan<char> format, Numb
         private static string FormatDouble(ref ValueStringBuilder sb, double value, ReadOnlySpan<char> format, NumberFormatInfo info)
         {
             char fmt = ParseFormatSpecifier(format, out int digits);
-            int precision = DoublePrecision;
+
             NumberBuffer number = default;
             number.kind = NumberBufferKind.Double;
 
-            switch (fmt)
+            if ((fmt == 'R') || (fmt == 'r'))
             {
-                case 'R':
-                case 'r':
-                    {
-                        // In order to give numbers that are both friendly to display and round-trippable, we parse the
-                        // number using 15 digits and then determine if it round trips to the same value. If it does, we
-                        // convert that NUMBER to a string, otherwise we reparse using 17 digits and display that.
-                        DoubleToNumber(value, DoublePrecision, ref number);
-                        if (number.scale == ScaleNAN)
-                        {
-                            return info.NaNSymbol;
-                        }
-                        else if (number.scale == ScaleINF)
-                        {
-                            return number.sign ? info.NegativeInfinitySymbol : info.PositiveInfinitySymbol;
-                        }
-
-                        if (NumberToDouble(ref number) == value)
-                        {
-                            NumberToString(ref sb, ref number, 'G', DoublePrecision, info);
-                        }
-                        else
-                        {
-                            DoubleToNumber(value, 17, ref number);
-                            NumberToString(ref sb, ref number, 'G', 17, info);
-                        }
-
-                        return null;
-                    }
-
-                case 'E':
-                case 'e':
-                    // Round values less than E14 to 15 digits
-                    if (digits > 14)
-                    {
-                        precision = 17;
-                    }
-                    break;
-
-                case 'G':
-                case 'g':
-                    // Round values less than G15 to 15 digits. G16 and G17 will not be touched.
-                    if (digits > 15)
-                    {
-                        precision = 17;
-                    }
-                    break;
+                // IEEE Roundtripping requires us to include at least `17` digits
+                fmt = 'G';
+                digits = DefaultDoublePrecision;
             }
+            else if (digits <= 0)
+            {
+                // This ensures that, for the default case, we return a string containing the old
+                // number of digits (15), which results in "prettier" numbers for most cases.
+                digits = DefaultDoubleDigits;
```

Comment:
Now, `(0.1).ToString()` will print `0.1` (as before). But requesting more than `15` digits will return the requested number of digits (up to `50`).

For example, previously `(0.1).ToString("G20")` returned `"0.10000000000000001"` (which is the same as `"G17"`), but now will return `"0.10000000000000000555"` (as is required by the IEEE spec).

As mentioned [here](https://github.com/dotnet/coreclr/pull/19905#discussion_r216892369). The internal buffer for `DoubleToNumber` is set to `50`, but the precision specifier supports up to `99`. So it might be worth resizing it so that any requested number of digits works.